resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -830,49 +972,166 @@ void add(DnsRecord r) {
+                    serverName.update(parent.newRedirectServerAddress(resolved), r.timeToLive());
+                    return;
+                }
+                serverName = serverName.next;
+            }
+        }
+
+        // No handle all AuthoritativeNameServer for which we had no ADDITIONAL record
+        void handleAll(DnsNameResolver parent, DnsCache cache) {
             AuthoritativeNameServer serverName = head;
 
             while (serverName != null) {
-                if (!serverName.removed && serverName.nsName.equalsIgnoreCase(nsName)) {
-                    serverName.removed = true;
-                    return serverName;
+                if (serverName.address == null) {
+                    // Try to resolve via cache
+
+                    List<? extends DnsCacheEntry> entries = cache.get(serverName.nsName, null);
+                    if (entries == null || entries.isEmpty()) {
+                        // These will be resolved on the fly if needed.
+                        serverName.address = InetSocketAddress.createUnresolved(
+                                serverName.nsName, DefaultDnsServerAddressStreamProvider.DNS_PORT);
+                    } else {
+                        // We have some entries in the cache.
+                        for (int i = 0; i < entries.size(); i++) {
+                            DnsCacheEntry entry = entries.get(i);
+                            InetAddress address = entry.address();
+                            if (address != null) {
+                                if (serverName.address != null) {
+                                    AuthoritativeNameServer server = new AuthoritativeNameServer(serverName);
+                                    server.next = serverName.next;
+                                    serverName.next = server;
+                                    serverName = server;
+
+                                    count++;
+                                }
+                                // Use a TTL of Long.MIN_VALUE as we don't want to cache the actual address later on as
+                                // we have no idea about how long the TTL should be. When we encounter Long.MIN_VALUE
+                                // later on we will cache the unresolved address.
+                                serverName.update(parent.newRedirectServerAddress(address), Long.MIN_VALUE);
+                            }
+                        }
+                    }
                 }
                 serverName = serverName.next;
             }
-            return null;
         }
 
-        int size() {
-            return count;
+        /**
+         * Returns {@code true} if empty, {@code false} otherwise.
+         */
+        boolean isEmpty() {
+            return count == 0;
+        }
+
+        /**
+         * Creates a new {@link List} which holds the {@link InetSocketAddress}es.
+         */
+        List<InetSocketAddress> addressList() {
```

Comment:
I think I made this better with my latest changes (I removed one iteration).

Reply:
Here is what I had in mind to avoid this additional iteration at the cost of potentially copy/resize of the resulting list. It depends upon the use case but *maybe* resizing vs *always* doing another iteration maybe a good tradeoff to start. We didn't even handle the "no-additionals" case before this, so not sure how common this is?

```diff
diff --git a/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java b/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java
index b4c83e4c67..069c97dccf 100644
--- a/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java
+++ b/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java
@@ -487,6 +487,12 @@ abstract class DnsResolveContext<T> {
         if (res.count(DnsSection.ANSWER) == 0) {
             AuthoritativeNameServerList serverNames = extractAuthoritativeNameServers(question.name(), res);
             if (serverNames != null) {
+                // Make a best guess for how many elements there will be in the redirect list. This amount is variable
+                // because authority entries with no corresponding additional entries may result in adding a variable
+                // amount of addresses to this list from the cache.
+                List<InetSocketAddress> redirectedNameServerAddresses =
+                        new ArrayList<InetSocketAddress>(serverNames.getNameServerCount());
+                AuthoritativeDnsServerCache authoritativeDnsServerCache = authoritativeDnsServerCache();
                 int additionalCount = res.count(DnsSection.ADDITIONAL);
 
                 for (int i = 0; i < additionalCount; i++) {
@@ -499,17 +505,17 @@ abstract class DnsResolveContext<T> {
 
                     // We may have multiple ADDITIONAL entries for the same nameserver name. For example one AAAA and
                     // one A record.
-                    serverNames.handleWithAdditional(parent, r, authoritativeDnsServerCache());
+                    serverNames.handleWithAdditional(parent, r, authoritativeDnsServerCache,
+                            redirectedNameServerAddresses);
                 }
 
                 // Process all unresolved nameservers as well.
-                serverNames.handleWithoutAdditionals(parent, resolveCache(), authoritativeDnsServerCache());
-
-                List<InetSocketAddress> addresses = serverNames.addressList();
+                serverNames.handleWithoutAdditionals(parent, resolveCache(), authoritativeDnsServerCache,
+                        redirectedNameServerAddresses);
 
                 // Give the user the chance to sort or filter the used servers for the query.
                 DnsServerAddressStream serverStream = parent.newRedirectDnsServerStream(
-                        question.name(), addresses);
+                        question.name(), redirectedNameServerAddresses);
 
                 if (serverStream != null) {
                     query(serverStream, 0, question,
@@ -920,10 +926,10 @@ abstract class DnsResolveContext<T> {
 
         private final String questionName;
 
-        // We not expect the linked-list to be very long so a double-linked-list is overkill.
+        // We do not expect the linked-list to be very long so a double-linked-list is overkill.
         private AuthoritativeNameServer head;
 
-        private int servers;
+        private int nameServerCount;
 
         AuthoritativeNameServerList(String questionName) {
             this.questionName = questionName.toLowerCase(Locale.US);
@@ -967,20 +973,29 @@ abstract class DnsResolveContext<T> {
             // We are only interested in preserving the nameservers which are the closest to our qName, so ensure
             // we drop servers that have a smaller dots count.
             if (head == null || head.dots < dots) {
-                servers = 1;
                 head = new AuthoritativeNameServer(dots, r.timeToLive(), recordName, domainName);
+                nameServerCount = 1;
             } else if (head.dots == dots) {
                 AuthoritativeNameServer serverName = head;
                 while (serverName.next != null) {
                     serverName = serverName.next;
                 }
                 serverName.next = new AuthoritativeNameServer(dots, r.timeToLive(), recordName, domainName);
-                servers++;
+                ++nameServerCount;
             }
         }
 
+        boolean isEmpty() {
+            return head == null;
+        }
+
+        int getNameServerCount() {
+            return nameServerCount;
+        }
+
         void handleWithAdditional(
-                DnsNameResolver parent, DnsRecord r, AuthoritativeDnsServerCache authoritativeCache) {
+                DnsNameResolver parent, DnsRecord r, AuthoritativeDnsServerCache authoritativeCache,
+                List<InetSocketAddress> redirectedNameServerAddresses) {
             // Just walk the linked-list and mark the entry as handled when matched.
             AuthoritativeNameServer serverName = head;
 
@@ -1003,8 +1018,7 @@ abstract class DnsResolveContext<T> {
                         server.next = serverName.next;
                         serverName.next = server;
                         serverName = server;
-
-                        servers++;
+                        ++nameServerCount; // optional, this isn't used after this point
                     }
                     // We should replace the TTL if needed with the one of the ADDITIONAL record so we use
                     // the smallest for caching.
@@ -1012,6 +1026,7 @@ abstract class DnsResolveContext<T> {
 
                     // Cache the server now.
                     cache(serverName, authoritativeCache, parent.executor());
+                    redirectedNameServerAddresses.add(serverName.address);
                     return;
                 }
                 serverName = serverName.next;
@@ -1020,17 +1035,16 @@ abstract class DnsResolveContext<T> {
 
         // Now handle all AuthoritativeNameServer for which we had no ADDITIONAL record
         void handleWithoutAdditionals(
-                DnsNameResolver parent, DnsCache cache, AuthoritativeDnsServerCache authoritativeCache) {
+                DnsNameResolver parent, DnsCache cache, AuthoritativeDnsServerCache authoritativeCache,
+                List<InetSocketAddress> redirectedNameServerAddresses) {
             AuthoritativeNameServer serverName = head;
 
             while (serverName != null) {
                 if (serverName.address == null) {
                     // These will be resolved on the fly if needed.
                     cacheUnresolved(serverName, authoritativeCache, parent.executor());
-                    servers++;
 
                     // Try to resolve via cache as we had no ADDITIONAL entry for the server.
-
                     List<? extends DnsCacheEntry> entries = cache.get(serverName.nsName, null);
                     if (entries != null && !entries.isEmpty()) {
                         InetAddress address = entries.get(0).address();
@@ -1038,6 +1052,7 @@ abstract class DnsResolveContext<T> {
                         // If address is null we have a resolution failure cached so just use an unresolved address.
                         if (address != null) {
                             serverName.update(parent.newRedirectServerAddress(address));
+                            redirectedNameServerAddresses.add(serverName.address);
 
                             for (int i = 1; i < entries.size(); i++) {
                                 address = entries.get(i).address();
@@ -1049,10 +1064,14 @@ abstract class DnsResolveContext<T> {
                                 serverName.next = server;
                                 serverName = server;
                                 serverName.update(parent.newRedirectServerAddress(address));
-
-                                servers++;
+                                ++nameServerCount; // optional, this isn't used after this point
+                                redirectedNameServerAddresses.add(serverName.address);
                             }
+                        } else {
+                            redirectedNameServerAddresses.add(serverName.address);
                         }
+                    } else {
+                        redirectedNameServerAddresses.add(serverName.address);
                     }
                 }
                 serverName = serverName.next;
@@ -1061,10 +1080,6 @@ abstract class DnsResolveContext<T> {
 
         private static void cacheUnresolved(
                 AuthoritativeNameServer server, AuthoritativeDnsServerCache authoritativeCache, EventLoop loop) {
-            // As we have no idea about how long the TTL should be we will just cache the unresolved
-            // address
-
-            // We still want to cached the unresolved address
             server.address = InetSocketAddress.createUnresolved(
                     server.nsName, DefaultDnsServerAddressStreamProvider.DNS_PORT);
 
@@ -1074,40 +1089,9 @@ abstract class DnsResolveContext<T> {
 
         private static void cache(AuthoritativeNameServer server, AuthoritativeDnsServerCache cache, EventLoop loop) {
             // Cache NS record if not for a root server as we should never cache for root servers.
-            if (!server.isRootServer() && server.address != null) {
-                // If we resolved the AuthoritativeNameServer via the DnsCache before we should not cache the
-                // resolved address as we don't have a good idea about the TTL to apply. We will just cache
-                // the unresolved address for now. We will then query the cache again the next time we try to
-                // use it for a query which will do the correct thing in terms of respecting the TTL of the
-                // A / AAAA record.
-                InetSocketAddress addressToCache = server.ttl == Long.MIN_VALUE ?
-                        InetSocketAddress.createUnresolved(server.nsName, server.address.getPort()) :
-                        server.address;
-                cache.cache(server.domainName, addressToCache, server.ttl, loop);
-            }
-        }
-
-        /**
-         * Returns {@code true} if empty, {@code false} otherwise.
-         */
-        boolean isEmpty() {
-            return servers == 0;
-        }
-
-        /**
-         * Creates a new {@link List} which holds the {@link InetSocketAddress}es.
-         */
-        List<InetSocketAddress> addressList() {
-            List<InetSocketAddress> addressList = new ArrayList<InetSocketAddress>(servers);
-
-            AuthoritativeNameServer server = head;
-            while (server != null) {
-                if (server.address != null) {
-                    addressList.add(server.address);
-                }
-                server = server.next;
+            if (!server.isRootServer()) {
+                cache.cache(server.domainName, server.address, server.ttl, loop);
             }
-            return addressList;
         }
     }
 ```

Reply:
So after talking with Scott we will keep it as it is as it often happens that the NS record count is not a good way to size the list.

For example it may see:

```
; <<>> DiG 9.10.6 <<>> @192.12.94.30 ns1.google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 13399
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 4, ADDITIONAL: 9
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;ns1.google.com.			IN	A

;; AUTHORITY SECTION:
google.com.		172800	IN	NS	ns2.google.com.
google.com.		172800	IN	NS	ns1.google.com.
google.com.		172800	IN	NS	ns3.google.com.
google.com.		172800	IN	NS	ns4.google.com.

;; ADDITIONAL SECTION:
ns2.google.com.		172800	IN	AAAA	2001:4860:4802:34::a
ns2.google.com.		172800	IN	A	216.239.34.10
ns1.google.com.		172800	IN	AAAA	2001:4860:4802:32::a
ns1.google.com.		172800	IN	A	216.239.32.10
ns3.google.com.		172800	IN	AAAA	2001:4860:4802:36::a
ns3.google.com.		172800	IN	A	216.239.36.10
ns4.google.com.		172800	IN	AAAA	2001:4860:4802:38::a
ns4.google.com.		172800	IN	A	216.239.38.10
```

Reply:
fine as is ... just for more context we can also account for the ADDITIONAL count up front too ... but accounting up front will only ever be an estimate and may require resize (tradeoff is *maybe* resize or over allocation vs *always* additional iteration).